platform-sdk/docs/proposals/queuestate-fcqueue-to-virtualmap/proposal.md

```diff
@@ -0,0 +1,159 @@
+  a string value
+* These two longs are the head and the tail of the queue - see below
+
+### Queue operations
+
+The underlying `VirtualMap` will map longs to E, where E is the type of the queue. The longs
+will be used to organize the links between queue elements: if an element has position N, then
+its next element has position N+1. When an empty queue is created, its head and tail are both
+pointing to 0. The head is _the next_ position to add, the tails is _the current_ position to
+remove. If head and tail positions are equal, the queue is empty.
+
+![Map keys](map-keys.svg)
+
+**Add**:
+
+* Insert an element to the head position
+* Increment the head by one
+* Code snippet:
+
+```java
+public void add(final E element) {
+    map.put(head++, element);
+}
+```
+
+![Add to the queue](queue-add.svg)
+
+**Poll / remove**:
+
+* Check if the queue is empty
+* Remove the element from the tail position
+* Increment the tail by one
+* Code snippet:
+
+```java
+public E remove() {
+    if (tail == head) {
+        return null;
+    }
+    return map.remove(tail++);
+}
+```
+
+![Remove from the queue](queue-remove.svg)
+
+### Serialization, deserialization, hashing, reconnects
+
+All provided by `VirtualMap`.
+
+`QueueNodeState` will write both its long values in addition to the label during serialization.
+
+### Data serialization and deserialization
+
+`FCQueue` today stores all data items in memory. When the queue is written to a snapshot, all
+items are serialized as `SelfSerializable`. To store items in `VirtualMap`, key and value
+serializers are to be provided. The key serializer will be trivial, as keys are longs. The value
+serializer will be implemented using the codec provided by `QueueNode`, similar to how key/value
+states work.
+
+## Data migration
+
+Current production states contain a few queue state nodes, all are tiny except transaction
+receipts. However, even the largest queue is expected to be small enough, so it can be migrated
+to a virtual map during Hedera node software update window.
+
+Migration code will be in `QueueNode.migrate()` This method will get its right child (which is
+`FCQueue`), iterate over it, insert all elements in the right order to a virtual map, then set
+the map as the new right child. The left child will be changed from a simple string leaf to a
+`QueueNodeState` leaf node.
+
+## Testing
+
+This feature doesn't add any new functionality. All existing Hedera unit tests should pass with
+the new queue state implementation. Unit tests specific to `QueueNode` may need to be updated.
+
+### Migration testing
+
+Migration testing will use a real mainnet state generated with a previous Hedera version. This
+state will be used to start a node with the latest version that includes queue state changes.
+
+## Performance impact
+
+**Queue operations (add, remove)**: both `FCQueue` and `VirtualMap` add and remove elements fast.
+No performance impact is expected.
```

Comment:
Could you please expand on the "fast" part? Specifically, I'd like to see the performance of all the operations expressed using the big-O notation for both the old and the new implementations.

Additionally, it would be helpful to provide sizes of existing queues in the current state as well as describe (/speculate about) their future growth to help understand if the new implementation's performance characteristics support the future growth.

Reply:
"Fast" here means O(1), both for FCQueue and VirtualMap.

Existing queue sizes: several queues with just a few (under a hundred, I guess elements) and one queue with about 1.8M elements (transaction receipts at 10K TPS). The latter shouldn't grow over time, it's to keep transaction receipts from the last 3 mins. The size of this queue is more or less constant, as older receipts are removed from the queue automatically

Reply:
Got it. Thanks. I'd guess the transaction receipts queue is going to grow as the network grows, but I agree it's not going to be an explosive growth.

Regardless, as long as all the operations are O(1) for both the implementations, I think we should be good as far as scaling is concerned.

I suggest to include the big-O and queue sizes in the document for future reference.